storage/client/internal/curl_wrappers.cc

```diff
@@ -31,19 +32,50 @@ extern "C" std::size_t WriteCallback(void* contents, std::size_t size,
   return size * nmemb;
 }
 
+extern "C" std::size_t HeaderCallback(char* contents, std::size_t size,
+                                      std::size_t nmemb, void* dest) {
+  auto* headers = reinterpret_cast<storage::internal::CurlHeaders*>(dest);
+  headers->Append(contents, size * nmemb);
+  return size * nmemb;
+}
+
 }  // anonymous namespace
 
 namespace storage {
 inline namespace STORAGE_CLIENT_NS {
 namespace internal {
-void CurlBuffer::CaptureOutputOf(CURL* curl) {
+void CurlBuffer::Attach(CURL* curl) {
   curl_easy_setopt(curl, CURLOPT_WRITEDATA, this);
   curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback);
 }
 
 void CurlBuffer::Append(char* data, std::size_t size) {
   buffer_.append(data, size);
 }
+
+void CurlHeaders::Attach(CURL* curl) {
+  curl_easy_setopt(curl, CURLOPT_HEADERDATA, this);
+  curl_easy_setopt(curl, CURLOPT_HEADERFUNCTION, HeaderCallback);
+}
+
+void CurlHeaders::Append(char* data, std::size_t size) {
+  if (size <= 2) {
+    // Empty header (including the \r\n), ignore.
+    return;
+  }
+  if (data[size - 1] != '\n' or data[size - 2] != '\r') {
+    // Invalid header (should end in \r\n), ignore.
+    return;
+  }
+  auto separator = std::find(data, data + size, ':');
```

Comment:
Should we check for the case where someone passes something silly that doesn't have a ":" character?

Reply:
We sort of do?  If the character is missing we capture the full line as a header without a value (the comment tries to explain this by saying "if there is a value, capture it".  Do you think we should take a different action (ignoring the row, logging, or raising an exception are all options here)?  Or maybe I should improve the comments?